tower-http/src/cors.rs

```diff
@@ -0,0 +1,749 @@
+#[derive(Clone, Debug)]
+pub struct Origin(OriginInner);
+
+impl Origin {
+    /// Set a single allow origin target
+    pub fn exact(origin: HeaderValue) -> Self {
+        Self(OriginInner::Exact(origin))
+    }
+
+    /// Set multiple allow origin targets
+    pub fn list<I>(origins: I) -> Self
+    where
+        I: IntoIterator<Item = HeaderValue>,
+    {
+        let origins = origins.into_iter().collect::<Vec<_>>().into();
+        Self(OriginInner::List(origins))
+    }
+}
+
+#[derive(Clone)]
+enum OriginInner {
+    Exact(HeaderValue),
+    List(Arc<[HeaderValue]>),
+    Closure(Arc<dyn for<'a> Fn(&'a HeaderValue, &'a Parts) -> bool + Send + Sync + 'static>),
+}
+
+impl fmt::Debug for OriginInner {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Exact(inner) => f.debug_tuple("Exact").field(inner).finish(),
+            Self::List(inner) => f.debug_tuple("List").field(inner).finish(),
+            Self::Closure(_) => f.debug_tuple("Closure").finish(),
+        }
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for Cors<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Default,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, ResBody>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let origin = req.headers().get(&header::ORIGIN).cloned();
+
+        let origin = if let Some(origin) = origin {
+            origin
+        } else {
+            // This is not a CORS request if there is no Origin header
+            return ResponseFuture {
+                inner: Kind::NonCorsCall(self.inner.call(req)),
+            };
+        };
+
+        let (parts, body) = req.into_parts();
+
+        if !self.is_valid_origin(&origin, &parts) {
+            return ResponseFuture {
+                inner: Kind::Error(Some(
+                    Response::builder()
+                        .status(StatusCode::UNAUTHORIZED)
+                        .body(ResBody::default())
+                        .unwrap(),
+                )),
+            };
+        }
+
+        let req = Request::from_parts(parts, body);
+
+        // Return results immediately upon preflight request
+        if req.method() == Method::OPTIONS {
+            // the method the real request will be made with
+            match req.headers().get(header::ACCESS_CONTROL_REQUEST_METHOD) {
+                Some(request_method) => {
+                    if !self.is_valid_request_method(request_method) {
+                        return ResponseFuture {
+                            inner: Kind::Error(Some(
+                                Response::builder()
+                                    .status(StatusCode::METHOD_NOT_ALLOWED)
```

Comment:
I’m not sure returning a `METHOD_NOT_ALLOWED` (or `UNAUTHORIZED` below) is what’s supposed to happen here. Because from the client POV it would look like the `OPTIONS` method was not allowed / authorized

if `Access-Control-Allow-Methods` is present and does not include the `ACCESS_CONTROL_REQUEST_METHOD` it looks like it’s enough to stop the CORS pre flight request : https://fetch.spec.whatwg.org/#ref-for-concept-request-method%E2%91%A0%E2%91%A6

Reply:
I see that `METHOD_NOT_ALLOWED` might be confusing. Actix returns `bad request` but their internal error is called `MethodNotAllowed` 🤷 Which status do you think would be more appropriate? I'm not a fan of just terminating the connection without returning a response.

Reply:
For a pre-fligh `OPTIONS` request, I would expect a `No Content` response, that includes  a (possibly empty) `Access-Control-Allow-Methods` header. 

In fact, if the `Access-Control-Allow-Methods` is always present (I did not check) and the list of authorized methods is static, I would not manually check for the `ACCESS_CONTROL_REQUEST_METHOD` header. 

So unless you plan to add a way to provide a callback to determine dynamically if a request method would be accepted, like with `allow_origin(|origin ...| ...)`

---

additional remarks:

* the current behavior seems appropriate when it’s not a pre-flight request but I’m not sure if this header is included in a non pre-flight request
* the `is_valid_origin` check looks okay for a non pre-flight request but should also return a 204 in pre-flight mode (with the `Access-Control-Allow-Origin` header containing or not the origin depending if it’s a valid origin

Reply:
I'm not sure if 405 is good for CORS.
I think generally Bad Request would be better.

Semantically 405 indicates that resource doesn't support method at all.
In general I think following is ok:
400 (as more generic might better) or 403 as CORS is kinda auth issue?

Reply:
I stll don’t think it should be a 4xx at all, in a preflight request, you’re asking "can I do X ?", this request is OK even if the answer is no. and to express "no", you _just_ have to send back the correct CORS headers

Reply:
Agreed with @mathroc here. Did some digging and it seems like 200 might be the better status code to return.

- There's a [neat summary on Stackoverflow](https://stackoverflow.com/a/46028619) about the Fetch spec requiring that an "ok status" be sent back even in invalid cases. The "ok status" is defined as 2xx in the spec, but I'm guess some browsers interpreted it narrowly as just 200. Therefore, 204 did cause problems with legacy browsers (found an [old Firefox bug](https://bugzilla.mozilla.org/show_bug.cgi?id=845881)) and apparently with Smart TVs as well.
- Express.js allows you to customize the response code for CORS through `[optionsSuccessStatus](https://expressjs.com/en/resources/middleware/cors.html#configuration-options)`. They also [set Content-Length to 0](https://github.com/expressjs/cors/blob/master/lib/index.js#L176-L180) and [default to 204 status](https://github.com/expressjs/cors/blob/c49ca10e92ac07f98a3b06783d3e6ba0ea5b70c7/lib/index.js#L12).
- I ran a test with actix-cors on my current server. It _is_ returning 400, but says "Requested method is not allowed" just as @davidpdrsn confirmed. This seems incorrect to me.

Reply:
There is nothing wrong with using 400.
From server perspective it is bad request.
I suppose you can make exception for options since it doesn't suppose to actually perform something on server so for `OPTIONS` it should be 204, not 200.
For a general case (i.e. any other method) it should be 400

Reply:
@dDoumanAsh, yes we're talking about preflight ( = `OPTIONS`) requests here

Reply:
Haha I'm glad you reached an agreement. I've changed the status to `OK`.